src/wp-includes/class-wp-block-type.php

```diff
@@ -540,4 +563,21 @@ public function get_attributes() {
 			$this->attributes :
 			array();
 	}
+
+	/**
+	 * Get block variations.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return array[]
+	 */
+	public function get_variations() {
+		if ( isset( $this->variation_callback ) && is_callable( $this->variation_callback )
+			&& ( empty( $this->variations ) || apply_filters( 'allow_variation_callback_recall', false ) )
+		) {
+				$this->variations = call_user_func( $this->variation_callback );
+		}
+
+		return $this->variations;
```

Comment:
Let's separate out the conditions to make it a bit more readable and to allow us to document any filters we add to this method. 

I don't think that we want to support re-running a variation callback via a filter, as that could cause manually set variations to be overwritten. Instead, I think we could filter the return value so that extenders that want to modify the registered variations can do so after the fact. In doing so, we could pass the original object so that if someone wanted to manually rerun the callback they could. Here's my suggestion:

```suggestion
		if ( empty( $this->variations && is_callable( $this->variation_callback ) ) ) {
				$this->variations = call_user_func( $this->variation_callback );
		}

		/**
		 * Filters the registered variations for a block type.
		 *
		 * @since 6.5.0
		 *
		 * @param array                    $variations   Array of registered variations for a block type.
		 * @param WP_Block_Type $block_type The full block type object.
		 */
		return apply_filters( 'get_block_type_variations', $this->variations, $this );
```

Reply:
What if the callback returns any empty array? That would mean it would run every time?

Reply:
@spacedmonkey A valid point. It doesn't make sense to run it every time if the callback returns an empty array. To avoid it I have now initialised `$variations` to null; setting it to an empty array only when it is referenced. Now, call to callback also happens only when `$variations` is not set.

@joemcgill  re-running a variation callback via a filter is risky, there is a risk of some plugin accidentally setting it to true, so updated the code the way you suggested.